protocol/message_persistence.go

```diff
@@ -2207,3 +2181,161 @@ func (db sqlitePersistence) SetContactRequestState(id string, state common.Conta
 	_, err := db.db.Exec(`UPDATE user_messages SET contact_request_state = ? WHERE id = ?`, state, id)
 	return err
 }
+
+func getUpdatedChatMessagePayload(originalMessage *protobuf.DiscordMessage, attachmentMessage *protobuf.DiscordMessage) *protobuf.ChatMessage_DiscordMessage {
+	originalMessage.Attachments = append(originalMessage.Attachments, attachmentMessage.Attachments...)
+	return &protobuf.ChatMessage_DiscordMessage{
+		DiscordMessage: originalMessage,
+	}
+}
+
+func getMessageFromScanRows(db sqlitePersistence, rows *sql.Rows) (*common.Message, error) {
+	var msg *common.Message
+
+	for rows.Next() {
+		// There's a possibility of multiple rows per message if the
+		// message has a discordMessage and the discordMessage has multiple
+		// attachments
+		//
+		// Hence, we make sure we're aggregating all attachments on a single
+		// common.Message
+		var message common.Message
+		err := db.tableUserMessagesScanAllFields(rows, &message)
+		if err != nil {
+			return nil, err
+		}
+
+		if msg == nil {
+			msg = &message
+		} else if discordMessage := msg.GetDiscordMessage(); discordMessage != nil {
+			msg.Payload = getUpdatedChatMessagePayload(discordMessage, message.GetDiscordMessage())
+		}
+	}
+	return msg, nil
+}
+
+type Messages []*common.Message
+type PinnedMessages []*common.PinnedMessage
+
+func (m Messages) Len() int            { return len(m) }
+func (m Messages) Swap(i, j int)       { m[i], m[j] = m[j], m[i] }
+func (m PinnedMessages) Len() int      { return len(m) }
+func (m PinnedMessages) Swap(i, j int) { m[i], m[j] = m[j], m[i] }
+
+type ByClock struct{ Messages }
+type ByClockPinned struct{ PinnedMessages }
+
+func (m ByClock) Less(i, j int) bool {
+	return m.Messages[i].Clock < m.Messages[j].Clock
+}
+
+func (m ByClockPinned) Less(i, j int) bool {
+	return m.PinnedMessages[i].Message.Clock < m.PinnedMessages[j].Message.Clock
+}
```

Comment:
Both `ByClock `and `ByClockPinned` look a bit over engineered to my yes.
I assume there's a nicer way to get sorting capabilities of `[]*someStructType` in go?

Reply:
You are basically doing the best you can here, because you are choosing which property of `Message` to sort by. It is annoying, but for structs you have multiple fields that a sorting algorithm has no way to default to. There is a way we can make this code a bit more reusable.

```go
type HasClocks interface {
    GetClock(i int) int
}

func (m *Messages) GetClock(i int) int {
    return m.Messages[i].Clock
}

func (m *PinnedMessages) GetClock(i int) int {
    return m.PinnedMessages[i].Message.Clock
}

func SortByClock(msgs HasClocks) {
    sort.Slice(msgs, func(i, j int) bool {
        return msgs.GetClock(j) < msgs.GetClock(i)
    })
}
```

Notice in the `sort.Slice` func I compare `j < i` this will allow for reverse sorting by default.
Also you no longer need to implement the `sort.Interface`

Reply:
Alright so, I gave this a try but couldn't get it to compile.
It seems it is required that we move the `Messages` and `PinnedMessages` types into a struct and implement `HasClocks` there, because go doesn't recognize `HasClocks` being implemented on these types otherwise.
Once moved into a struct, the sort.Slice doesn't seem to work anymore.

So for now I've rolled back to what I had before, renamed `ByClock` and `ByClockPinned` to `MessagesWrapper` and `PinnedMessagesWrapper` and moved them to `common`.

Reply:
As discussed offline, I made a mistake with the type annotations.
This is no implemented.